### PR TITLE
A simple fake News classifier implemented on idris

### PR DESCRIPTION
**Pull request by Eman22S**
This branch contains an implementation of Neural Networks, activation functions, loss functions specifically, Softmax activation layer, Cross entropy loss function and Adam optimization done from scratch in Idris. There is also a haskel version to this classifier not pushed. Testing on the the accuracy of prediction as well as conversion of csv files directly using idris is not yet implemented.

### CODE REVIEW

**Comment by ngeiswei** [comment]
Very cool, @Eman22S!

I'll carefully review that tomorrow, meanwhile could you
1. move that code to `experimental/ai-algorithms`
2. remove the following files
```
experimental/NeuralNets/*.ibc
experimental/NeuralNets/build/*
experimental/NeuralNets/util.idr
```
followed by clean force push?  Thanks.

**Comment by Eman22S** [comment]
@ngeiswei Yes definitely we don't need those build files

**Comment by ngeiswei** [comment]
@Eman22S, could you
1. rebase your branch onto the master,
2. move your `NeuralNets` folder under `experimental/ai-algorithms`,
3. then push to another branch than `master`,
4. close that PR and recreate a new one with that branch?
Thanks.

File: experimental/NeuralNets/Lib.idr [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+export
+nums : String
+nums = "0123456789!@#$%^&*()_+=-:;"
+
+export
+removeStopWords : String -> List String
+removeStopWords line = do
+              let tmp = words line
+              foldr (\x, acc => if elem x stopwords then acc else x::acc ) [] tmp
```
Comment (ngeiswei) [inline experimental/NeuralNets/Lib.idr:16]: @Eman22S, I'm not sure the usage of `do` and `let` is generally appropriate for code without side effect like this one.  Using `where` or something instead might be more appropriate/idiomatic (@sroberti might have thoughts on this).

Also, you can simplify that line by using `filter` or list comprehension instead of `foldr`.  If you use `filter` you can also probably avoid using lambda by combining `flip` and partial evaluation.

File: experimental/NeuralNets/Lib.idr [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+export
+makeWordPairs : List String -> List (String, String)
+makeWordPairs [] = []
+makeWordPairs (x::y::z::h::xs) = (x,y)::(x,z)::(y,z)::(y,h)::(z,x)::(z,y)::(z,h)::(h,x)::(h,y)::(h,z)::makeWordPairs xs
+makeWordPairs (x::y::z::xs) = (x,y)::(x,z)::(y,z)::(z,x)::(z,y)::makeWordPairs xs
+makeWordPairs (x::y::xs) = (x,y)::makeWordPairs xs
```
Comment (ngeiswei) [inline experimental/NeuralNets/Lib.idr:23]: Comments are welcome as soon as functions are not trivial, this one is kinda on the edge.  It looks like it's making pairs of between neighboring words, not completely sure.

File: experimental/NeuralNets/Lib.idr [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+adam_Optimizer (t::thetas) (gt::gts) alpha m_prev v_prev beta1 beta2 epsilon = do
+              let m_current = (beta1 * m_prev + (1-beta1) * gt)/ (1-beta1)
+              let v_current = (beta2 * v_prev + (1-beta2) *  (db.pow gt 2))/(1-(db.pow beta2 2))
+              let mhat_current = m_current /(1-beta1)
+              let vhat_current = v_current/(1-beta2)
+              (t - alpha * mhat_current / ((sqrt  vhat_current) + epsilon)):: adam_Optimizer thetas gts alpha m_current v_current beta1 beta2 epsilon
+
+export
+wordpairs : List String -> List (String,String)
+wordpairs wrds = makeWordPairs wrds
```
Comment (ngeiswei) [inline experimental/NeuralNets/Lib.idr:87]: You don't need `wrds`, you can just write

```idris
wordpairs = makeWordPairs
```

File: experimental/NeuralNets/Lib.idr [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+export
+fullNNLayer : List (List Double) -> List (List Double) -> List (List Double)
+fullNNLayer [] [] = []
+fullNNLayer (x::inputs) (y::weights) = (softmax_activation $ nn_Layer x y)::fullNNLayer inputs weights
+
+export
+loss_func : (List Double -> List Double -> List Double ) -> List (List Double) -> List (List Double) -> List (List Double)
+loss_func f [] _ = []
+loss_func f _ [] = []
+loss_func f [] [] = []
```
Comment (ngeiswei) [inline experimental/NeuralNets/Lib.idr:126]: I think that pattern `loss_func f [] [] = []` is already covered above.

File: experimental/NeuralNets/Lib.idr [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+loss_func f (p::preds) (t::targets) = f p t :: loss_func f preds targets
+
+export
+optimizer : ( List Double -> List Double -> Double -> Double -> Double -> Double -> Double -> Double -> List Double) -> List (List Double) -> List (List Double) -> Double -> Double -> Double -> Double -> Double -> Double -> List (List Double)
+optimizer f [] _ _ _ _ _ _ _ = []
+optimizer f _ [] _ _ _ _ _ _ = []
+optimizer f [] [] _ _ _ _ _ _= []
```
Comment (ngeiswei) [inline experimental/NeuralNets/Lib.idr:133]: I think that pattern `optimizer f [] [] = []` is already covered above.

File: experimental/NeuralNets/Lib.idr [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+optimizer f (ws::wss) (cs::costs) alpha  m v beta1 beta2 epsilon =  (f ws cs alpha m v beta1 beta2 epsilon) :: optimizer f wss costs alpha m  v beta1 beta2 epsilon
+
+export
+model : List (List Double) -> List (List Double) -> List (List Double) -> Double -> Double -> Double -> Double -> Int -> List (List Double )
+model [] _ _ _ _ _ _ _ = []
+model _ [] _ _ _ _ _ _ = []
+model _ _ [] _ _ _ _ _ = []
+model [] [] [] _ _ _ _ _= []
```
Comment (ngeiswei) [inline experimental/NeuralNets/Lib.idr:141]: I think that pattern `model [] [] [] ... = []` is already covered above, but please correct me if I'm wrong.

**Review by ngeiswei — COMMENTED** [review]
@Eman22S, could add a couple lines of comment for each non trivial function of your code, it will make easier to review.

BTW, it looks like idris has 2 ways to comment
1. `--`
2. `|||`
I'm not sure but `|||` looks like it is for some sort of docstring, while `--` is more for regular comment (do you know something about that @sroberti ?).

**Comment by Eman22S** [comment]
@Nil thank you for your reviews I will try to work on them and send a new PR  after a rebase.

File: experimental/NeuralNets/Lib.idr (hunk 1 above)
Comment (Eman22S) [reply experimental/NeuralNets/Lib.idr:16]: Yes. `Do` can be replaced by `let ..in`. The initial implementation was using `filter` . I used `foldr` here and other places since `foldr` can minimize the cost of using `filter` and `reverse` later on. since inserting elements using `:` reverses the order of initial order.  `filter ` and `reverse` are both `O(n)` .  Also by partial evaluation , do you mean something like curried functions? I tried using them without much lack. Clearly it's different from how haskell uses curried functions.(Idris requires type declaration where as haskell do not) @ sroberti please share your thoughts if you have any ideas.

File: experimental/NeuralNets/Lib.idr (hunk 2 above)
Comment (Eman22S) [reply experimental/NeuralNets/Lib.idr:23]: Will add comments here.

File: experimental/NeuralNets/Lib.idr (hunk 4 above)
Comment (Eman22S) [reply experimental/NeuralNets/Lib.idr:126]: Good point since `_` also covers `[ ]`.

File: experimental/NeuralNets/Lib.idr (hunk 3 above)
Comment (Eman22S) [reply experimental/NeuralNets/Lib.idr:87]: Correct!

File: experimental/NeuralNets/Lib.idr (hunk 1 above)
Comment (ngeiswei) [reply experimental/NeuralNets/Lib.idr:16]: I don't understand, are you saying that `filter` would be slower than your use of `foldr`?

Looking at its implementation I don't believe it would (but I didn't try)
https://github.com/idris-lang/Idris2/blob/299a31de5bd64a584a2f7d9090eb2f5680efaeac/libs/base/Data/List.idr#L107-L112

Comment (ngeiswei) [reply experimental/NeuralNets/Lib.idr:16]: What I had in mind was
```idris
removeStopWords line = filter ((flip elem) stopwords) (words line)
```
Wouldn't that be semantically equivalent and as fast?

**Comment by ngeiswei** [comment]
Comment syntax, `--`, `|||`, etc, is described here

http://docs.idris-lang.org/en/latest/reference/documenting.html

It's for idris1 but is still valid for idris2.

File: experimental/NeuralNets/Lib.idr (hunk 1 above)
Comment (Eman22S) [reply experimental/NeuralNets/Lib.idr:16]: Sorry for not responding sooner ..holy month has been getting me busy lately anyways. I think I am going with the mine because yours return simply the stop words which is not what we want. removestopwords should return the text weeding out stop words for example. The prince is the Future king will be [prince, future, king]. Yours return [the, is] which we don't want. I have removed the do clause however as it is indeed unnecessary

Comment (ngeiswei) [reply experimental/NeuralNets/Lib.idr:16]: Oh right, it would be something like
```idris
removeStopWords line = filter (not . ((flip elem) stopwords)) (words line)
```
then.

Comment (ngeiswei) [reply experimental/NeuralNets/Lib.idr:16]: For sake of completion here's the version using `filter` and `lambda`
```idris
removeStopWords line = filter (\x => not (elem x stopwords)) (words line)
```
as well as the version using list comprehension
```
removeStopWords line = [x | x <- (words line), not (elem x stopwords))]
```
